unit-tests-java8/src/test/java/org/eclipse/collections/test/RichIterableUniqueTestCase.java

```diff
@@ -678,6 +679,43 @@ default void RichIterable_groupBy()
         assertSame(target2, actualWithTarget);
     }
 
+    @Override
+    @Test
+    default void RichIterable_groupBy2()
+    {
+        RichIterable<Integer> iterable = this.newWith(4, 3, 2, 1);
+        Function<Integer, Boolean> isOdd = object -> IntegerPredicates.isOdd().accept(object);
+        Function<Integer, Boolean> inRange = object -> Predicates.betweenInclusive(1, 2).accept(object);
+
+        Map<Boolean, Multimap<Boolean, Integer>> expectedGroupBy =
+                UnifiedMap.newWithKeysValues(
+                        Boolean.TRUE, this.newWith(2, 1).groupBy(isOdd),
+                        Boolean.FALSE, this.newWith(4, 3).groupBy(isOdd));
+
+        assertEquals(expectedGroupBy, iterable.groupBy2(inRange, isOdd));
+    }
+
+    @Override
+    @Test
+    default void RichIterable_groupBy3()
+    {
+        RichIterable<Integer> iterable = this.newWith(4, 3, 2, 1, -1, -2, -3, -4);
+        Function<Integer, Boolean> isPositive = object -> IntegerPredicates.isPositive().accept(object);
+        Function<Integer, Boolean> isOdd = object -> IntegerPredicates.isOdd().accept(object);
+        Function<Integer, Boolean> inRange = object -> Predicates.betweenInclusive(-2, 2).accept(object);
+
+        Map<Boolean, Map<Boolean, Multimap<Boolean, Integer>>> expectedGroupBy =
+                UnifiedMap.newWithKeysValues(
+                        Boolean.TRUE, UnifiedMap.newWithKeysValues(
+                            Boolean.TRUE, this.newWith(2, 1).groupBy(isOdd),
+                            Boolean.FALSE, this.newWith(4, 3).groupBy(isOdd)),
+                        Boolean.FALSE,  UnifiedMap.newWithKeysValues(
+                            Boolean.TRUE, this.newWith(-2, -1).groupBy(isOdd),
+                            Boolean.FALSE, this.newWith(-4, -3).groupBy(isOdd)));
+
+        assertEquals(expectedGroupBy, iterable.groupBy3(isPositive, inRange, isOdd));
+    }
```

Comment:
This should work for a 3 level groupBy:

`iterable.reduceInPlace(Collectors.groupingBy(inRange, Collectors2.toListMultimap(isOdd)))`

The reason I point this out is that it might be useful for testing.  The underlying call in toListMultimap calls a groupBy call with a Supplier.  It is however harder to make utility calls like this covariant as you have with your new groupBy3 implementation.